### DISCUSSION TITLE
`TypeGuard` is assignable to `bool`

### DISCUSSION THREAD

**Discussion by layday** [microsoft/pyright#2404] [General]
[PEP 647](https://www.python.org/dev/peps/pep-0647/#typeguard-type), under the heading "TypeGuard Type", says:

> In all other respects, `TypeGuard` is a distinct type from bool. It is not a subtype of bool. Therefore, `Callable[..., TypeGuard[int]]` is not assignable to `Callable[..., bool]`.

However, 1. type checks both in Pyright and mypy, and both seem to treat `TypeGuard` as a subtype of `bool`. This means that the overload order in 2. is significant:

```py
# pyright: strict

from typing import Callable, Iterable, TypeVar, overload
from typing_extensions import TypeGuard


# 1.

def foo(guard: Callable[..., bool]) -> None:
    ...

def bar(value: object) -> TypeGuard[int]:
    return True


foo(bar)

baz: Callable[..., bool] = bar


# 2.

_T = TypeVar('_T')
_S = TypeVar('_S')

@overload
def qux(guard: Callable[[_T], bool], value: Iterable[_T]) -> Iterable[object]: ...
@overload
def qux(guard: Callable[[_T], TypeGuard[_S]], value: Iterable[_T]) -> Iterable[_S]: ...

quux = list(qux(bar, ['a']))
reveal_type(quux)  # Type of "quux" is "list[object]"

```

Am I misinterpreting the PEP, is this uninentional, or is it intentional and does the PEP need updating?

**Comment by erictraut** [comment]
You are misinterpreting the PEP. You would need to swap the two overloads to make this work. The `TypeGuard` should come first.
